common.blocks/i-bem-dom/__event-manager/_type/i-bem-dom__event-manager_type_bem.js

```diff
@@ -0,0 +1,231 @@
+                    blockName = defCls._blockName;
+                    elemName = ctx;
+                } else if(typeOfCtx === 'object') { // bem entity with optional mod val
+                    blockName = ctx.block?
+                        ctx.block.getName() :
+                        defCls._blockName;
+                    elemName = typeof ctx.elem === 'function'?
+                        ctx.elem.getName() :
+                        ctx.elem;
+                    modName = ctx.modName;
+                    modVal = ctx.modVal;
+                } else if(ctx.getName() === ctx.getEntityName()) { // block class
+                    blockName = ctx.getName();
+                } else { // elem class
+                    blockName = defCls._blockName;
+                    elemName = ctx.getName();
+                }
+
+                entityName = BEMINTERNAL.buildClass(blockName, elemName);
+                res.entityCls = getEntityCls(entityName);
+                res.selector = '.' + (res.key = entityName + BEMINTERNAL.buildModPostfix(modName, modVal));
+            }
+        } else {
+            res.entityCls = defCls;
+        }
+
+        return res;
+    },
+
+    EventManager = inherit({
+        __constructor : function(params, fnWrapper) {
+            this._params = params;
+            this._fnWrapper = fnWrapper;
+            this._storage = {};
+        },
+
+        /**
+         * Adds an event handler
+         * @param {String} e Event type
+         * @param {Object} [data] Additional data that the handler gets as e.data
+         * @param {Function} fn Handler
+         * @returns {EventManager} this
+         */
+        on : function(e, data, fn) {
+            if(typeof e === 'object') {
+                objects.each(e, function(fn, e) {
+                    this.on(e, data, fn);
+                }, this);
+            } else {
+                var params = this._params;
+
+                e = buildEventName(e, params.entityCls.getEntityName());
+
+                if(arguments.length === 2) {
+                    fn = data;
+                    data = undef;
+                }
+
+                var fnStorage = this._storage[e] || (this._storage[e] = {}),
+                    fnId = identify(fn);
+                params.ctx.on(
+                    e,
+                    params.selector,
+                    data,
+                    fnStorage[fnId] = this._fnWrapper(fn, fnId));
+            }
+
+            return this;
+        },
+
+        /**
+         * Removes event handler or handlers
+         * @param {String} [e] Event type
+         * @param {Function} [fn] Handler
+         * @returns {EventManager} this
+         */
+        un : function(e, fn) {
+            if(typeof e === 'object') {
+                objects.each(e, function(fn, e) {
+                    this.un(e, fn);
+                }, this);
+            } else {
+                var argsLen = arguments.length;
+                if(!argsLen) {
```

Comment:
Also, `if(!condition) { /* some */ } else { /* other */ }` could be also rewritten as `if(condition) { /* other */ } else { /* some */ }`.

Or (better) with early returns.